### PR TITLE
Pause/Resume Upload for Transfer Manager with Java S3Client

### PR DESCRIPTION
**Pull request by davidh44**
## Motivation and Context

Currently pausing/resuming a file upload with TransferManager is only supported with CRT S3 Client.

## Modifications

Adding support when using Java S3 Client.
building off PR https://github.com/aws/aws-sdk-java-v2/pull/4886

## Testing

added integ tests
to add unit tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadWithKnownContentLengthHelper.java [hunk 1]
```diff
@@ -205,20 +313,32 @@ public void onSubscribe(Subscription s) {
 
         @Override
         public void onNext(AsyncRequestBody asyncRequestBody) {
-            log.trace(() -> "Received asyncRequestBody " + asyncRequestBody.contentLength());
+            if (isPaused) {
+                return;
+            }
+
+            if (listPartsFuture != null && !listPartsFuture.isDone()) {
+                listPartsFuture.join();
```
Comment (zoewangg) [inline services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadWithKnownContentLengthHelper.java:321]: As discussed offline, we should chain future instead of doing .join

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/SdkPojoConversionUtils.java [hunk 2]
```diff
@@ -107,6 +109,18 @@ public static CompletedPart toCompletedPart(UploadPartResponse partResponse, int
         return builder.partNumber(partNumber).build();
     }
 
+    public static CompletedPart toCompletedPart(Part part) {
+        CompletedPart.Builder builder = CompletedPart.builder();
+        setSdkFields(builder, part);
+        return builder.build();
+    }
+
+    public static ListPartsRequest toListPartsRequest(String uploadId, PutObjectRequest putObjectRequest) {
```
Comment (zoewangg) [inline services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/SdkPojoConversionUtils.java:118]: Can we add tests?

File: services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/S3TransferManagerUploadPauseAndResumeTest.java [hunk 3]
```diff
@@ -129,18 +167,29 @@ void resumeUploadFile_fileModified_shouldAbortExistingAndUploadFromBeginning() {
...
 
         AbortMultipartUploadRequest actualRequest = abortMultipartUploadRequestArgumentCaptor.getValue();
         assertThat(actualRequest.uploadId()).isEqualTo(multipartId);
     }
 
-    @Test
-    void resumeUploadFile_hasValidResumeToken_shouldResumeUpload() {
+    @ParameterizedTest
+    @MethodSource("transferManagers")
+    void resumeUploadFile_hasValidResumeToken_shouldResumeUpload(TmType tmType) {
```
Comment (zoewangg) [inline services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/S3TransferManagerUploadPauseAndResumeTest.java:192]: Can we verify ListParts request as well for Java TM?

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadWithKnownContentLengthHelper.java [hunk 4]
```diff
@@ -44,10 +53,10 @@
 public final class UploadWithKnownContentLengthHelper {
     private static final Logger log = Logger.loggerFor(UploadWithKnownContentLengthHelper.class);
 
+    private static final ConcurrentHashMap<Integer, CompletedPart> EXISTING_PARTS = new ConcurrentHashMap<>();
```
Comment (cenedhryn) [inline services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadWithKnownContentLengthHelper.java:56]: I've been looking for tests that test the different code paths in this class, but there is no test class for this class. There are TM functional tests, but I think it would be valuable to have direct unit tests of this class now that it's pretty complex - not sure why they were not included from the beginning. We can discuss offline.

**Review by cenedhryn — APPROVED** [review]

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/PauseObservable.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.s3.internal.multipart;
+
+import software.amazon.awssdk.annotations.SdkInternalApi;
+
+@SdkInternalApi
```
[the remark this answers is not in the document]
Comment (davidh44) [reply services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/PauseObservable.java:20]: moved this and others to s3/multipart module

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadWithKnownContentLengthHelper.java [hunk 6]
```diff
@@ -205,20 +313,32 @@ public void onSubscribe(Subscription s) {
...
         @Override
         public void onNext(AsyncRequestBody asyncRequestBody) {
-            log.trace(() -> "Received asyncRequestBody " + asyncRequestBody.contentLength());
+            if (isPaused) {
+                return;
+            }
+
+            if (listPartsFuture != null && !listPartsFuture.isDone()) {
+                listPartsFuture.join();
+            }
```
[the remark this answers is not in the document]
Comment (davidh44) [reply services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadWithKnownContentLengthHelper.java:322]: refactored to not use join(), so not needed anymore

**Review by cenedhryn — APPROVED** [review]

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadWithKnownContentLengthHelper.java (hunk 4 above)
Comment (davidh44) [reply services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadWithKnownContentLengthHelper.java]: this class is covered by `UploadObjectHelperTest`, added a test there and also created a test class for `KnownContentLengthAsyncRequestBodySubscriber`

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/MpuRequestContext.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+@SdkInternalApi
+public class MpuRequestContext {
+    private final Pair<PutObjectRequest, AsyncRequestBody> request;
+    private final long contentLength;
+    private final long partSize;
+    private final long numPartsCompleted;
+    private final String uploadId;
+    private final Map<Integer, CompletedPart> existingParts;
+
+    protected MpuRequestContext(Pair<PutObjectRequest, AsyncRequestBody> request,
```
Comment (zoewangg) [inline services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/MpuRequestContext.java:35]: Nit: let's use builder pattern. Since this is an internal class, we don't need to create an interface for builder

File: .changes/next-release/feature-AmazonS3-a20b910.json [hunk 8]
```diff
@@ -0,0 +1,6 @@
+{
+    "category": "Amazon S3", 
+    "contributor": "", 
+    "type": "feature", 
+    "description": "Add support for pause/resume upload for TransferManager with Java-based S3 Client"
```
Comment (zoewangg) [inline .changes/next-release/feature-AmazonS3-a20b910.json:5]: nit, with Java-based S3 client that has multipart enabled.

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/KnownContentLengthAsyncRequestBodySubscriber.java [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+import software.amazon.awssdk.services.s3.model.CompletedPart;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectResponse;
+import software.amazon.awssdk.services.s3.model.UploadPartRequest;
+import software.amazon.awssdk.services.s3.multipart.S3ResumeToken;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Pair;
+
+@SdkInternalApi
+public class KnownContentLengthAsyncRequestBodySubscriber implements Subscriber<AsyncRequestBody>  {
```
Comment (zoewangg) [inline services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/KnownContentLengthAsyncRequestBodySubscriber.java:41]: Thanks for extracting it out, can we create separate commit or separate PR for refactoring and making changes because it's really hard to see the actual code change.

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/multipart/S3ResumeToken.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+        private Builder() {
+        }
+
+        public Builder uploadId(String uploadId) {
+            this.uploadId = uploadId;
+            return this;
+        }
+
+        public Builder partSize(long partSize) {
```
Comment (zoewangg) [inline services/s3/src/main/java/software/amazon/awssdk/services/s3/multipart/S3ResumeToken.java:89]: We should use boxed type in builder and add validation in the ctor

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadWithKnownContentLengthHelper.java [hunk 11]
```diff
@@ -99,160 +111,128 @@ private void uploadInParts(PutObjectRequest putObjectRequest, long contentLength
...
-                             .toArray(CompletedPart[]::new);
-                multipartUploadHelper.completeMultipartUpload(returnFuture, uploadId, parts, putObjectRequest);
-            }
+        public S3ResumeToken pause() {
+            return subscriber.pause();
         }
+    }
 
+    @SdkTestInternalApi
+    public static DefaultPausableUpload defaultPausableUpload(KnownContentLengthAsyncRequestBodySubscriber subscriber) {
```
Comment (zoewangg) [inline services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadWithKnownContentLengthHelper.java:235]: This is a bit code smell; we should try to find other non-intrusive way to mock dependencies such as dependency injection.

I see this method is used in `DefaultFileUploadTest`. Do we actually need to replace subscriber? Can we mock `PausableUpload` and verify the mocked pausable upload?

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/multipart/PauseObservable.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    public S3ResumeToken pause() {
+        // single part upload or TM is not used
+        if (pausableUpload == null) {
+            return null;
+        }
+        return pausableUpload.pause();
+    }
+
+    @SdkTestInternalApi
+    public boolean pausableUploadSet() {
```
Comment (zoewangg) [inline services/s3/src/main/java/software/amazon/awssdk/services/s3/multipart/PauseObservable.java:40]: Maybe we can make it getter method, `PausableUpload pausableUpload()`? Seems legit to add a getter and we can remove `SdkTestInternalApi`

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/KnownContentLengthAsyncRequestBodySubscriber.java [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+            CompletedPart completedPart = existingParts.containsKey(currPart) ? existingParts.get(currPart) :
+                                          completedParts.get(currPart);
+            merged[currPart - 1] = completedPart;
+            currPart++;
+        }
+        return merged;
+    }
+
+    @SdkTestInternalApi
+    public void setCompleteMpuFuture(CompletableFuture<CompleteMultipartUploadResponse> completeMpuFuture) {
```
Comment (zoewangg) [inline services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/KnownContentLengthAsyncRequestBodySubscriber.java:216]: Same here, suggesting using dependency injection for testing instead (passing a mock `multipartUploadHelper` that returns a mocked completeMpuFuture)

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/KnownContentLengthAsyncRequestBodySubscriber.java [hunk 14]
```diff
@@ -201,7 +201,7 @@ private void completeMultipartUploadIfFinished(int requestsInFlight) {
     }
 
     private CompletedPart[] mergeCompletedParts() {
-        CompletedPart[] merged = new CompletedPart[partCount];
+        CompletedPart[] merged = new CompletedPart[(int) partCount];
```
Comment (zoewangg) [inline services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/KnownContentLengthAsyncRequestBodySubscriber.java:203]: This seems dangerous. Looks like max part number is 10,000 and I don't think it'll ever exceed max integer value... Should we change the type to integer then?

Comment (davidh44) [reply services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/KnownContentLengthAsyncRequestBodySubscriber.java]: updated `partCount` and `numExistingParts` to int

**Review by zoewangg — APPROVED** [review]
Good job!
